### PR TITLE
feat: add support for JSON data type

### PR DESCRIPTION
**Pull request by olavloite**
Adds support for `JSON` data type.

### CODE REVIEW

File: samples/json-update-data.js
Comment (olavloite) [inline samples/json-update-data.js:53]: @zoercai Here's an example of the issue with top-level JSON arrays. The client library does not know whether this should be encoded as `JSON` or `ARRAY<JSON>` if it is specified as an actual array instead of a string. Specifying it as a string will cause the client library to just encode it as a string, which is the default encoding for `JSON` values.

File: src/codec.ts
Comment (olavloite) [inline src/codec.ts:427]: This is where the most important part of this PR is: It ensures that any value that has been specified as an object will be converted to a valid `JSON` string. An array of objects will be converted to an array of valid `JSON` strings.

**Review by skuruppu — APPROVED** [review]

File: src/transaction.ts
Comment (olavloite) [inline src/transaction.ts:1564]: @skuruppu This adds additional information to the error for the specific case that the application tries to insert an array of objects into a JSON column.

Comment (skuruppu) [reply src/transaction.ts:1564]: Perfect, thanks for doing this. If you could also please add some unit tests for this, that would be awesome.

Comment (olavloite) [reply src/transaction.ts:1564]: Good point, added.

**Review by skuruppu — APPROVED** [review]

File: src/transaction.ts
Comment (skuruppu) [reply src/transaction.ts:1564]: Thanks a lot for adding the tests.

**Comment by skuruppu** [comment]
@olavloite I just merged in the codegen PR so you can rebase this PR now.

**Comment by olavloite** [comment]
I added the Do not merge label back as the system test is failing with

```
  1) Spanner
       types
         "before all" hook for "should throw an error for incorrect value types":
     Error: 12 UNIMPLEMENTED: JSON type is not enabled.
      at Object.callErrorFromStatus (node_modules/@grpc/grpc-js/build/src/call.js:31:26)
          -> /workspace/node_modules/@grpc/grpc-js/src/call.ts:81:24
      at Object.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/client.js:179:52)
          -> /workspace/node_modules/@grpc/grpc-js/src/client.ts:338:36
      at /workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:80:35
          -> /workspace/node_modules/@grpc/grpc-js/src/call-stream.ts:170:27
      at Object.onReceiveStatus (node_modules/grpc-gcp/build/src/index.js:73:29)
          -> /workspace/node_modules/grpc-gcp/src/index.ts:129:15
      at InterceptingListenerImpl.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/call-stream.js:75:23)
          -> /workspace/node_modules/@grpc/grpc-js/src/call-stream.ts:166:19
      at Object.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/client-interceptors.js:336:141)
          -> /workspace/node_modules/@grpc/grpc-js/src/client-interceptors.ts:426:34
      at Object.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/client-interceptors.js:299:181)
          -> /workspace/node_modules/@grpc/grpc-js/src/client-interceptors.ts:389:48
      at /workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:145:78
          -> /workspace/node_modules/@grpc/grpc-js/src/call-stream.ts:276:24
      at processTicksAndRejections (internal/process/task_queues.js:79:11)
```

File: test/codec.ts
Comment (skuruppu) [inline test/codec.ts:688]: Should we also add a case for decoding a deeply-nested object or is that overkill?

Comment (olavloite) [reply test/codec.ts:688]: I think it's reasonable to have it symmetrical, so I've added a test for that as well.

File: test/codec.ts
Comment (skuruppu) [inline test/codec.ts:707]:
```suggestion
      // This test ensures that the decoder is able to decode such a string.
```

**Review by skuruppu — APPROVED** [review]

File: test/codec.ts
Comment (olavloite) [reply test/codec.ts]: Done

**Comment by zoercai** [comment]
Hi @olavloite, we're getting ready to merge the JSON PRs now for Node, Java, and Python. You can merge this PR as soon as all the tests pass. Could you please rebase and rerun CI? Thanks!

**Comment by olavloite** [comment]
> Hi @olavloite, we're getting ready to merge the JSON PRs now for Node, Java, and Python. You can merge this PR as soon as all the tests pass. Could you please rebase and rerun CI? Thanks!

The JSON samples and other JSON tests pass, but there is some other error occurring for the system tests at the moment that blocks merging this. That error is not specific to this PR, but happens for multiple (all?) PRs at the moment.

**Comment by skuruppu** [comment]
I'm reaching out to the Node.js team to figure out why this suddently started happening.

**Comment by skuruppu** [comment]
@olavloite if you can please take a look at the test failure, that would be great. I ran it a couple of times and got the same error so I gathered it must be due to the new tests we added in this PR.

`ERROR: Error: 6 ALREADY_EXISTS: Row [81] in table Venues already exists`.

**Comment by olavloite** [comment]
> @olavloite if you can please take a look at the test failure, that would be great. I ran it a couple of times and got the same error so I gathered it must be due to the new tests we added in this PR.
> 
> `ERROR: Error: 6 ALREADY_EXISTS: Row [81] in table Venues already exists`.

It was caused by a merge/rebase error. The previously added test runs for the tagging feature were also added once more by this PR, which meant that the tagging sample tests were executed twice. I've removed the duplicate test runs, and hopefully that should fix it.

**Comment by skuruppu** [comment]
Now asked for help from the Node.js team on how to unblock the OwlBot check.

**Pull request merged**
